internal/server/tls.go

```diff
@@ -17,20 +16,51 @@ import (
 	"github.com/infrahq/infra/internal/logging"
 )
 
+// MapCache is a simple in-memory caching mechanism, it is not thread safe
+type MapCache map[string][]byte
+
+func (m MapCache) Get(ctx context.Context, name string) ([]byte, error) {
+	data, cached := m[name]
+	if !cached {
+		return nil, autocert.ErrCacheMiss
+	}
+	return data, nil
+}
+
+func (m MapCache) Put(ctx context.Context, name string, data []byte) error {
+	m[name] = data
+	return nil
+}
+
+func (m MapCache) Delete(ctx context.Context, name string) error {
+	delete(m, name)
+	return nil
+}
+
+func (m MapCache) getKeyPair(ctx context.Context, serverName string) (cert, key []byte) {
+	certBytes, _ := m.Get(ctx, serverName+".crt")
+	keyBytes, _ := m.Get(ctx, serverName+".key")
+	if certBytes == nil || keyBytes == nil {
+		logging.Infof("no cached TLS cert for %v", serverName)
+	}
+	return certBytes, keyBytes
+}
+
+var certCache = make(MapCache)
```

Comment:
Made these package level rather than function level variables so we can just share them and simplify the logic of `getCertificate`

Reply:
We really need to avoid the use of package level variables (#1939).

I think keeping the closure that we had before removes the need for this package level variable.